### ISSUE TITLE
Unable to send request with body larger than 64Kb

### ISSUE THREAD

**Issue by GRoguelon** [sneako/finch#265]
Hi,

I am not totally sure if this is a bug but I found a stranged behaviour in [Req](https://github.com/wojtekmach/req/issues/309) when I'm sending a body larger than 64Kb and it seems to be related to Finch `protocols` option.

It seems a change in the `protocols` provides different behaviours:

With `protocols: [:http1]`:
```elixir
iex(1)> {:ok, pid} = Finch.start_link(name: MyFinch, pools: %{ default: [ protocols: [:http1] ] })
{:ok, #PID<0.648.0>}
iex(2)> Finch.build(:post, "https://nghttp2.org/httpbin/post", [], :crypto.strong_rand_bytes(65538)) |> Finch.request(MyFinch)
{:ok,
 %Finch.Response{
   status: 200,
   body: "{\n  \"args\": {}, \n  \"data\": \"data:application/octet-stream;base64,TRUNCATED",
   headers: [
     {"date", "Tue, 20 Feb 2024 18:31:19 GMT"},
     {"content-type", "application/json"},
     {"content-length", "87690"},
     {"access-control-allow-origin", "*"},
     {"access-control-allow-credentials", "true"},
     {"x-backend-header-rtt", "0.156761"},
     {"strict-transport-security", "max-age=31536000"},
     {"connection", "close"},
     {"alt-svc", "h3=\":443\"; ma=3600, h3-29=\":443\"; ma=3600"},
     {"server", "nghttpx"},
     {"via", "1.1 nghttpx"},
     {"x-frame-options", "SAMEORIGIN"},
     {"x-xss-protection", "1; mode=block"},
     {"x-content-type-options", "nosniff"}
   ],
   trailers: []
 }}
```

With `protocols: [:http2]`:
```elixir
iex(1)> {:ok, pid} = Finch.start_link(name: MyFinch, pools: %{ default: [ protocols: [:http2] ] })
{:ok, #PID<0.648.0>}
iex(2)> Finch.build(:post, "https://nghttp2.org/httpbin/post", [], :crypto.strong_rand_bytes(65538)) |> Finch.request(MyFinch)
{:ok,
 %Finch.Response{
   status: 200,
   body: "{\n  \"args\": {}, \n  \"data\": \"data:application/octet-stream;base64,TRUNCATED",
   headers: [
     {"date", "Tue, 20 Feb 2024 18:27:43 GMT"},
     {"content-type", "application/json"},
     {"content-length", "87695"},
     {"access-control-allow-origin", "*"},
     {"access-control-allow-credentials", "true"},
     {"x-backend-header-rtt", "0.508807"},
     {"strict-transport-security", "max-age=31536000"},
     {"server", "nghttpx"},
     {"alt-svc", "h3=\":443\"; ma=3600, h3-29=\":443\"; ma=3600"},
     {"via", "1.1 nghttpx"},
     {"x-frame-options", "SAMEORIGIN"},
     {"x-xss-protection", "1; mode=block"},
     {"x-content-type-options", "nosniff"}
   ],
   trailers: []
 }}
```

With `protocols: [:http1, :http2]`:
```elixir
iex(1)> {:ok, pid} = Finch.start_link(name: MyFinch, pools: %{ default: [ protocols: [:http1, :http2] ] })
{:ok, #PID<0.648.0>}
iex(2)> Finch.build(:post, "https://nghttp2.org/httpbin/post", [], :crypto.strong_rand_bytes(65538)) |> Finch.request(MyFinch)
{:error,
 %Mint.HTTPError{
   reason: {:exceeds_window_size, :request, 65535},
   module: Mint.HTTP2
 }}
```

Is it related to the lack of [multiplexing](https://hexdocs.pm/finch/Finch.html#start_link/1-pool-configuration-options) when `:http1` is provided?

Regards,

**Comment by wojtekmach** [comment]
Briefly looking at Finch.HTTP2.Pool, it implements flow control as described in https://hexdocs.pm/mint/Mint.HTTP2.html#get_window_size/2. The `protocols: [:http1, :http2]` option, talking HTTP/2 over HTTP/1 connection (Finch.HTTP1.Conn) does _not_ as it's using just the Mint.HTTP API.

We might have to port HTTP/2 flow control in HTTP/1 connections or give up on the http1/http2 alpn feature. I can look into this.

**Comment by ericmj** [comment]
You can check if it's a HTTP/2 connection after ALPN upgrade is done and call the `Mint.HTTP2` functions directly.

**Comment by whatyouhide** [comment]
So ok let me back up a sec to understand this better. Mint itself is behaving correctly it seems, as it's yelling when `protocols: [:http2]` and it's not yelling when `protocols: [:http1]`. So far so good.

Finch, instead, is behaving incorrectly because the issue only shows up with `protocols: [:http1, :http2]`, but **not** with `protocols: [:http2]` (where it *should* show up). Is that correct @sneako @wojtekmach?

What's happening with `protocols: [:http2]` that makes this *not show up* in Finch?

Sorry for backtracking here a bit, but I don't have that much time to dig into Finch's codebase right now and I’m trying to help out with this as best as I can 😄

**Comment by sneako** [comment]
No worries Andrea, I appreciate your attention here.

My understanding is that this issue pops up when we end up using Finch's HTTP1 pool but the connections in that pool are using HTTP2 due to ALPN

**Comment by whatyouhide** [comment]
Aaah okay, I see. One thing I still don't understand: we use `protocols: [:http1, :http2]` with Finch, so we use the HTTP/1 pool but the connections have upgraded to HTTP/2 through ALPN. 

Okay, but why does the Finch HTTP/2 pool handle large req payloads correctly? If the Finch HTTP/2 pool uses HTTP/2 conns, does it do anything to payloads to prevent this error?

**Comment by sneako** [comment]
That's because our H2 pool implements flow control as Wojtek pointed out above

**Comment by wojtekmach** [comment]
So to be more specific it's perhaps because H2 always seems to set request body to `:stream`, https://github.com/sneako/finch/blob/v0.18.0/lib/finch/http2/pool.ex#L547, and perhaps then Mint does all of the bookkeeping under the hood after all?

**Comment by neilberkman** [comment]
I ran into this issue while working on [req_llm](https://github.com/agentjido/req_llm) when sending a large prompt to an LLM API. When request bodies exceed 64KB with `protocols: [:http1, :http2]`, we hit the `{:exceeds_window_size, :request, 65535}` error.

**Workaround:** We addressed this in req_llm by switching to `protocols: [:http1]` which works reliably with all servers. See https://github.com/agentjido/req_llm/pull/113 for the implementation.

**Test case:** I've opened PR #332 with a comprehensive test that reproduces the bug. The failing test is marked with `@tag :skip` so it's available for future work without breaking CI. Two control tests confirm HTTP/2-only and HTTP/1-only work correctly with large bodies.

**Attempted fix:** I had Claude Code attempt to fix this and found that implementing HTTP/2 flow control in the HTTP/1 pool or transferring upgraded connections to the HTTP/2 pool is infeasible without major refactoring. The HTTP/1 pool uses NimblePool with multiple workers, while HTTP/2 uses a single gen_statem for multiplexing. Safely transferring connection state between these would require fundamental changes to Finch's pool management.

I thought it was worth contributing the test since I had a failing test that replicated the issue, in case the maintainers decide to tackle this themselves.

**Comment by wojtekmach** [comment]
Yeah, at the moment we need to either do `protocols: [:http1]` OR `protocols: [:http2]`.
